app/models/salt_event.rb

```diff
@@ -0,0 +1,48 @@
+class SaltEvent < ApplicationRecord
+  # Processing of a single event shouldn't take more than than 5 minutes.
+  # We consider events which were taken 5 minutes ago but not yet processed as
+  # if the worker died so we reassign them for processing.
+  PROCESS_TIMEOUT_MIN = 5
+  NO_EVENTS_TIMEOUT_SEC = 5
+
+  def self.process(worker_id:)
+    logger.info "Salt event processor #{worker_id} started at #{Time.current}"
+
+    loop do
+      sleep(NO_EVENTS_TIMEOUT_SEC) if !process_next_event(worker_id: worker_id)
+    end
+  end
+
+  def self.process_next_event(worker_id:)
+    taken_event = SaltEvent.where(
+      "processed_at IS NULL AND (worker_id IS NULL OR (worker_id = ?) OR taken_at < ?)",
+      worker_id, Time.current - PROCESS_TIMEOUT_MIN.minutes)
+    .limit(1).update_all(worker_id: worker_id, taken_at: Time.current)
+
+    return false if taken_event == 0
+
+    SaltEvent.where(worker_id: worker_id, processed_at: nil).
+      where("taken_at IS NOT NULL").limit(1).first.process!
+  end
+
+  def handler
+    return @handler if @handler
+
+    case tag
+    when "minion_start"
+      @handler = SaltEventHandlers::MinionStart.new(self)
+    else
+      nil
+    end
+  end
+
+  def process!
+    handler.process_event if handler
+
+    update_column(:processed_at, Time.current)
+  end
+
+  def parsed_data
```

Comment:
I'm not sure. The overhead is small and I'm afraid that if we ever change the data attribute (don't have a case yet) then parsed_data will return the old version leading to bugs. If we need to call parsed_data multiple times we should either way store it in a variable because in any case we would be invoking a method multiple times instead of accessing a variable.